File: crates/video-coach-core/src/denormalize.rs
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct PixelPoint {
    pub x: f64,
    pub y: f64,
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub struct CanvasSize {
    pub width: f64,
    pub height: f64,
}

/// Map a normalized (top-left origin, x and y in 0..1) stroke point into pixels.
///
/// `flip_y = true` for live overlays in a bottom-left-origin coordinate space.
/// `flip_y = false` for the export compositor, which already applies its own flip.
/// See `docs/plans/2026-04-27-video-coach-design.md` § "Drawing capture".
pub fn point(x: f64, y: f64, into: CanvasSize, flip_y: bool) -> PixelPoint {
    let px = x * into.width;
    let py = y * into.height;
    if flip_y {
        PixelPoint {
            x: px,
            y: into.height - py,
        }
    } else {
        PixelPoint { x: px, y: py }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn no_flip_passes_through() {
        let p = point(
            0.5,
            0.5,
            CanvasSize {
                width: 1920.0,
                height: 1080.0,
            },
            false,
        );
        assert_eq!(p, PixelPoint { x: 960.0, y: 540.0 });
    }

    #[test]
    fn flip_inverts_y() {
        let p = point(
            0.0,
            0.0,
            CanvasSize {
                width: 1920.0,
                height: 1080.0,
            },
            true,
        );
        assert_eq!(p, PixelPoint { x: 0.0, y: 1080.0 });
    }
}

File: apple/VideoCoachCore/Sources/VideoCoachCore/Denormalize.swift
import CoreGraphics

/// Maps a stored normalized stroke point (top-left origin, x and y in `0...1`)
/// into a pixel position inside an output canvas of `size`.
///
/// `flipY` is the CRITICAL knob — call sites differ in whether they're drawing
/// into a top-left-origin or bottom-left-origin space:
///
/// - **Live overlay** (`NSView` with `isFlipped == false`, bottom-left origin):
///   pass `flipY: true` so the stored top-left y becomes a bottom-left y.
/// - **Export compositor** (`CGContext` over `CVPixelBuffer` with the
///   `translateBy(0, h); scaleBy(1, -1)` flip already applied): pass
///   `flipY: false`, since user-space is already top-left after the flip.
///
/// See `docs/plans/2026-04-27-video-coach-design.md` § "Drawing capture" for
/// the misuse table; passing `flipY: true` in the export compositor renders
/// strokes upside-down (a known footgun the smoke test guards against).
public enum Denormalize {
    public static func point(_ x: Double, _ y: Double, into size: CGSize, flipY: Bool) -> CGPoint {
        let px = CGFloat(x) * size.width
        let py = CGFloat(y) * size.height
        return flipY ? CGPoint(x: px, y: size.height - py) : CGPoint(x: px, y: py)
    }
}
